File: DFS_BFS/LC55-jumpGameBFS.py
from collections import deque
class Solution:
    def canJump(self, nums):
        
        n=len(nums)
        seen=set()
        q=deque([0])
        while q:
            cur=q.popleft()
            for i in range(0,nums[cur]+1):
                next=cur+i
                if next not in seen and next<n:
                    if next+nums[next]>=n-1:
                        return True;
                    q.append(next)
                    seen.add(next)
        return False
    
num=[3,2,1,0,4]
s=Solution()
print(s.canJump(num))

File: DFS_BFS/LC55-jumpGameBFS.cpp
#include<iostream>
#include<vector>
#include<deque>
#include<set>
using VEC=std::vector<int>;


bool canJump(const VEC& nums)
{
    int n=nums.size();
    if(n==1)
        return true;
    std::deque<int> q{0}; //store the indexes
    std::set<int> seen{};
    while(!q.empty())
    {
        int cur=q.front();
        q.pop_front();
        for(int i=0;i<=nums[cur];++i)
        {
            int next=i+cur;
            if(next<n&&seen.find(next)==seen.end())
            {
                if((nums[next]+next)>=n-1)
                    return true;

                q.push_back(next);
                seen.insert(next);
            }

        }
    }
    return false;
}

int main()
{
    VEC nums{3,2,4,0,0};
    std::cout<<canJump(nums);
}